File: notification_app_be/stage6.py
"""
Stage 6: Top 10 notification ranking with an O(n log 10) min-heap.

Run:
    python3 notification_app_be/stage6.py
"""

from __future__ import annotations

import heapq
import json
import logging
import os
import sys
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Dict, Iterable, List, Optional, Tuple
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


API_URL = "http://4.224.186.213/evaluation-service/notifications"
TOP_K = 10
TYPE_WEIGHTS = {
    "Placement": 3,
    "Result": 2,
    "Event": 1,
}
TIMESTAMP_FORMAT = "%Y-%m-%d %H:%M:%S"


class AffordmedLogger:
    """
    Lightweight Affordmed logging adapter.

    The local repository only contains a JS middleware package manifest, not a
    callable Python logging client. This adapter always logs to the console and
    optionally posts logs to AFFORDMED_LOG_URL when configured.
    """

    def __init__(self) -> None:
        logging.basicConfig(
            level=logging.INFO,
            format="[%(levelname)s] %(message)s",
        )
        self._logger = logging.getLogger("stage6")
        self._remote_url = os.getenv("AFFORDMED_LOG_URL")
        self._token = os.getenv("AFFORDMED_LOG_TOKEN")

    def info(self, message: str, **context: Any) -> None:
        self._write("info", message, context)

    def error(self, message: str, **context: Any) -> None:
        self._write("error", message, context)

    def _write(self, level: str, message: str, context: Dict[str, Any]) -> None:
        suffix = f" | {json.dumps(context, default=str)}" if context else ""
        if level == "error":
            self._logger.error("%s%s", message, suffix)
        else:
            self._logger.info("%s%s", message, suffix)

        if not self._remote_url:
            return

        payload = {
            "stack": "backend",
            "level": level,
            "package": "notification_app_be",
            "message": message if not context else f"{message} {context}",
        }
        headers = {"Content-Type": "application/json"}
        if self._token:
            headers["Authorization"] = f"Bearer {self._token}"

        try:
            request = Request(
                self._remote_url,
                data=json.dumps(payload).encode("utf-8"),
                headers=headers,
                method="POST",
            )
            with urlopen(request, timeout=3):
                pass
        except Exception as exc:  # Logging failures must not break the program.
            self._logger.debug("Remote Affordmed log failed: %s", exc)


logger = AffordmedLogger()


@dataclass(frozen=True)
class Notification:
    id: str
    type: str
    message: str
    timestamp_text: str
    timestamp: datetime

    @property
    def priority_score(self) -> Tuple[int, float]:
        """
        Heap priority.

        The type weight is first, so every Placement outranks every Result and
        every Result outranks every Event. The Unix timestamp is second, so newer
        notifications outrank older notifications within the same type.
        """
        return (TYPE_WEIGHTS[self.type], self.timestamp.timestamp())

    @property
    def display_score(self) -> str:
        weight, epoch = self.priority_score
        return f"{weight}.{int(epoch)}"


class TopNotificationTracker:
    """Maintains only the best K notifications in a min-heap."""

    def __init__(self, limit: int = TOP_K) -> None:
        self.limit = limit
        self._heap: List[Tuple[Tuple[int, float], int, Notification]] = []
        self._sequence = 0

    def update_top_notifications(self, new_notification: Notification) -> None:
        """
        Update Top K in O(log K).

        The heap root is the weakest current Top K item. New items are pushed
        until the heap reaches K; after that, heapreplace only runs when the new
        notification outranks the weakest current item.
        """
        self._sequence += 1
        entry = (new_notification.priority_score, self._sequence, new_notification)
        logger.info(
            "Priority calculation",
            notification_id=new_notification.id,
            notification_type=new_notification.type,
            priority_score=new_notification.display_score,
        )

        if len(self._heap) < self.limit:
            heapq.heappush(self._heap, entry)
            return

        if entry > self._heap[0]:
            heapq.heapreplace(self._heap, entry)

    def top_notifications(self) -> List[Notification]:
        logger.info("Top 10 generation", heap_size=len(self._heap))
        return [
            entry[2]
            for entry in sorted(
                self._heap,
                key=lambda item: item[0],
                reverse=True,
            )
        ]


tracker = TopNotificationTracker()


def update_top_notifications(new_notification: Dict[str, Any] | Notification) -> None:
    """
    Reusable update function for continuous arrivals.

    Accepts either the raw API-style dictionary or a validated Notification.
    """
    notification = (
        new_notification
        if isinstance(new_notification, Notification)
        else parse_notification(new_notification)
    )
    if notification is not None:
        tracker.update_top_notifications(notification)


def fetch_notifications() -> List[Dict[str, Any]]:
    logger.info("API request start", url=API_URL)

    try:
        request = Request(API_URL, headers={"Accept": "application/json"})
        with urlopen(request, timeout=10) as response:
            status = getattr(response, "status", 200)
            raw_body = response.read().decode("utf-8")
    except HTTPError as exc:
        logger.error("API failure", status=exc.code, reason=exc.reason)
        return []
    except URLError as exc:
        logger.error("API failure", reason=str(exc.reason))
        return []
    except TimeoutError as exc:
        logger.error("API failure", reason=str(exc))
        return []

    logger.info("API request success", status=status)

    try:
        payload = json.loads(raw_body)
    except json.JSONDecodeError as exc:
        logger.error("Invalid JSON", error=str(exc))
        return []

    notifications = payload.get("notifications") if isinstance(payload, dict) else None
    if notifications is None:
        logger.error("Missing notifications field")
        return []
    if not isinstance(notifications, list):
        logger.error("Invalid notifications field", expected="list")
        return []
    if not notifications:
        logger.error("Empty response", notification_count=0)
        return []

    logger.info("Notification count", count=len(notifications))
    return notifications


def parse_notification(raw: Dict[str, Any]) -> Optional[Notification]:
    required_fields = ("ID", "Type", "Message", "Timestamp")
    if not isinstance(raw, dict):
        logger.error("Invalid notification item", item=raw)
        return None

    missing = [field for field in required_fields if field not in raw]
    if missing:
        logger.error("Missing fields", missing=missing, item=raw)
        return None

    notification_type = str(raw["Type"])
    if notification_type not in TYPE_WEIGHTS:
        logger.error("Invalid notification type", notification_type=notification_type)
        return None

    timestamp_text = str(raw["Timestamp"])
    try:
        timestamp = datetime.strptime(timestamp_text, TIMESTAMP_FORMAT)
    except ValueError:
        logger.error("Invalid timestamp", timestamp=timestamp_text)
        return None

    return Notification(
        id=str(raw["ID"]),
        type=notification_type,
        message=str(raw["Message"]),
        timestamp_text=timestamp_text,
        timestamp=timestamp,
    )


def build_top_notifications(raw_notifications: Iterable[Dict[str, Any]]) -> List[Notification]:
    for raw_notification in raw_notifications:
        update_top_notifications(raw_notification)
    return tracker.top_notifications()


def print_top_notifications(notifications: List[Notification]) -> None:
    if not notifications:
        print("\nNo valid notifications available.")
        return

    print("\nTop 10 Notifications")
    print("-" * 118)
    print(
        f"{'Rank':<6}"
        f"{'ID':<18}"
        f"{'Type':<12}"
        f"{'Timestamp':<22}"
        f"{'Priority Score':<18}"
        "Message"
    )
    print("-" * 118)

    for rank, notification in enumerate(notifications, start=1):
        message = notification.message.replace("\n", " ")
        if len(message) > 42:
            message = f"{message[:39]}..."
        print(
            f"{rank:<6}"
            f"{notification.id:<18.18}"
            f"{notification.type:<12}"
            f"{notification.timestamp_text:<22}"
            f"{notification.display_score:<18}"
            f"{message}"
        )


def main() -> int:
    raw_notifications = fetch_notifications()
    top_notifications = build_top_notifications(raw_notifications)
    print_top_notifications(top_notifications)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: notification_app_be/stage6.js
"use strict";

const { Log } = require("../logging_middleware");

const API_URL =
  process.env.NOTIFICATION_API_URL ||
  "http://4.224.186.213/evaluation-service/notifications";

const TOP_K = Number(process.env.TOP_K || 10);
const TYPE_WEIGHTS = {
  Placement: 3,
  Result: 2,
  Event: 1,
};

async function safeLog(stack, level, packageName, message) {
  try {
    await Log(stack, level, packageName, message);
  } catch {
    // The assessment disallows console/inbuilt logger for app logs. Logging
    // failures are intentionally swallowed so the main program still runs.
  }
}

class MinHeap {
  constructor(compare) {
    this.items = [];
    this.compare = compare;
  }

  size() {
    return this.items.length;
  }

  peek() {
    return this.items[0];
  }

  push(item) {
    this.items.push(item);
    this.bubbleUp(this.items.length - 1);
  }

  replaceRoot(item) {
    this.items[0] = item;
    this.bubbleDown(0);
  }

  toArray() {
    return [...this.items];
  }

  bubbleUp(index) {
    while (index > 0) {
      const parent = Math.floor((index - 1) / 2);
      if (this.compare(this.items[index], this.items[parent]) >= 0) {
        break;
      }
      [this.items[index], this.items[parent]] = [
        this.items[parent],
        this.items[index],
      ];
      index = parent;
    }
  }

  bubbleDown(index) {
    const length = this.items.length;

    while (true) {
      const left = index * 2 + 1;
      const right = index * 2 + 2;
      let smallest = index;

      if (
        left < length &&
        this.compare(this.items[left], this.items[smallest]) < 0
      ) {
        smallest = left;
      }

      if (
        right < length &&
        this.compare(this.items[right], this.items[smallest]) < 0
      ) {
        smallest = right;
      }

      if (smallest === index) {
        break;
      }

      [this.items[index], this.items[smallest]] = [
        this.items[smallest],
        this.items[index],
      ];
      index = smallest;
    }
  }
}

function parseTimestamp(timestamp) {
  if (typeof timestamp !== "string") {
    return null;
  }

  const parsed = new Date(timestamp.replace(" ", "T"));
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

function comparePriority(a, b) {
  if (a.typeWeight !== b.typeWeight) {
    return a.typeWeight - b.typeWeight;
  }

  if (a.timestampMs !== b.timestampMs) {
    return a.timestampMs - b.timestampMs;
  }

  return a.sequence - b.sequence;
}

function validateNotification(raw, sequence) {
  if (!raw || typeof raw !== "object") {
    return { error: "notification item is not an object" };
  }

  for (const field of ["ID", "Type", "Message", "Timestamp"]) {
    if (!(field in raw)) {
      return { error: `missing field ${field}` };
    }
  }

  if (!(raw.Type in TYPE_WEIGHTS)) {
    return { error: `invalid notification type ${raw.Type}` };
  }

  const timestamp = parseTimestamp(raw.Timestamp);
  if (!timestamp) {
    return { error: `invalid timestamp ${raw.Timestamp}` };
  }

  return {
    notification: {
      id: String(raw.ID),
      type: raw.Type,
      message: String(raw.Message),
      timestampText: String(raw.Timestamp),
      timestampMs: timestamp.getTime(),
      typeWeight: TYPE_WEIGHTS[raw.Type],
      sequence,
    },
  };
}

class TopNotificationTracker {
  constructor(limit = TOP_K) {
    this.limit = limit;
    this.sequence = 0;
    this.heap = new MinHeap(comparePriority);
  }

  async updateTopNotifications(rawNotification) {
    this.sequence += 1;
    const { notification, error } = validateNotification(
      rawNotification,
      this.sequence
    );

    if (error) {
      await safeLog("backend", "error", "service", error);
      return;
    }

    await safeLog(
      "backend",
      "debug",
      "service",
      `priority calculated for ${notification.id}: ${notification.typeWeight}.${notification.timestampMs}`
    );

    if (this.heap.size() < this.limit) {
      this.heap.push(notification);
      return;
    }

    if (comparePriority(notification, this.heap.peek()) > 0) {
      this.heap.replaceRoot(notification);
    }
  }

  topNotifications() {
    return this.heap.toArray().sort((a, b) => comparePriority(b, a));
  }
}

const tracker = new TopNotificationTracker();

async function update_top_notifications(newNotification) {
  await tracker.updateTopNotifications(newNotification);
}

async function fetchNotifications() {
  await safeLog("backend", "info", "service", "notification API request start");

  try {
    const controller = new AbortController();
    const timeout = setTimeout(
      () => controller.abort(),
      Number(process.env.NOTIFICATION_TIMEOUT_MS || 10000)
    );

    const response = await fetch(API_URL, {
      method: "GET",
      headers: {
        Accept: "application/json",
        Authorization: `Bearer ${process.env.ACCESS_TOKEN || ""}`,
      },
      signal: controller.signal,
    });
    clearTimeout(timeout);

    await safeLog(
      "backend",
      "info",
      "service",
      `notification API request success with status ${response.status}`
    );

    if (!response.ok) {
      await safeLog(
        "backend",
        "error",
        "service",
        `notification API failure: status ${response.status}`
      );
      return [];
    }

    let data;
    try {
      data = await response.json();
    } catch {
      await safeLog("backend", "error", "service", "invalid JSON from API");
      return [];
    }

    if (!data || !Array.isArray(data.notifications)) {
      await safeLog("backend", "error", "service", "invalid notifications JSON");
      return [];
    }

    if (data.notifications.length === 0) {
      await safeLog("backend", "warn", "service", "empty notification response");
      return [];
    }

    await safeLog(
      "backend",
      "info",
      "service",
      `notification count ${data.notifications.length}`
    );

    return data.notifications;
  } catch (error) {
    await safeLog(
      "backend",
      "error",
      "service",
      `notification API failure: ${error.name || error.code || error.message}`
    );
    return [];
  }
}

function priorityScore(notification) {
  return `${notification.typeWeight}.${notification.timestampMs}`;
}

function printTopNotifications(notifications) {
  if (notifications.length === 0) {
    process.stdout.write("\nNo valid notifications available.\n");
    return;
  }

  process.stdout.write("\nTop 10 Priority Notifications\n");
  process.stdout.write("Priority rule: Placement > Result > Event, then newer first\n\n");
  process.stdout.write(
    `${"Rank".padEnd(6)}${"ID".padEnd(38)}${"Type".padEnd(12)}${"Timestamp".padEnd(22)}${"Priority Score".padEnd(22)}Message\n`
  );
  process.stdout.write("-".repeat(122) + "\n");

  notifications.forEach((notification, index) => {
    const message =
      notification.message.length > 42
        ? `${notification.message.slice(0, 39)}...`
        : notification.message;

    process.stdout.write(
      `${String(index + 1).padEnd(6)}` +
        `${notification.id.padEnd(38)}` +
        `${notification.type.padEnd(12)}` +
        `${notification.timestampText.padEnd(22)}` +
        `${priorityScore(notification).padEnd(22)}` +
        `${message}\n`
    );
  });
}

async function main() {
  const notifications = await fetchNotifications();

  for (const notification of notifications) {
    await update_top_notifications(notification);
  }

  const topNotifications = tracker.topNotifications();
  await safeLog(
    "backend",
    "info",
    "service",
    `top 10 generated with ${topNotifications.length} notifications`
  );
  printTopNotifications(topNotifications);
}

if (require.main === module) {
  main().catch(async (error) => {
    await safeLog("backend", "fatal", "service", `stage6 crashed: ${error.message}`);
    process.exitCode = 1;
  });
}

module.exports = {
  TopNotificationTracker,
  update_top_notifications,
  validateNotification,
};
